File: algorithms/15_ruby/data_structures/max_heap.rb
# Binary max-heap.
def push_heap(a, value)
  a << value
  i = a.length - 1
  while i > 0
    p = (i - 1) / 2
    break if a[p] >= a[i]

    a[p], a[i] = a[i], a[p]
    i = p
  end
end

def pop_heap(a)
  top = a[0]
  last = a.pop
  unless a.empty?
    a[0] = last
    i = 0
    loop do
      l = 2 * i + 1
      r = 2 * i + 2
      b = i
      b = l if l < a.length && a[l] > a[b]
      b = r if r < a.length && a[r] > a[b]
      break if b == i

      a[i], a[b] = a[b], a[i]
      i = b
    end
  end
  top
end

heap = []
[5, 3, 8, 1, 4].each { |v| push_heap(heap, v) }
previous = Float::INFINITY
until heap.empty?
  x = pop_heap(heap)
  raise "not a max-heap order" if x > previous

  previous = x
end
puts "max heap ok"

File: algorithms/13_swift/data_structures/max_heap.swift
// Binary max-heap.
func pushHeap(_ a: inout [Int], _ v: Int) {
    a.append(v)
    var i = a.count - 1
    while i > 0 {
        let p = (i - 1) / 2
        if a[p] >= a[i] { break }
        a.swapAt(p, i)
        i = p
    }
}

func popHeap(_ a: inout [Int]) -> Int {
    let top = a[0]
    let last = a.removeLast()
    if !a.isEmpty {
        a[0] = last
        var i = 0
        while true {
            let l = 2 * i + 1
            let r = 2 * i + 2
            var b = i
            if l < a.count && a[l] > a[b] { b = l }
            if r < a.count && a[r] > a[b] { b = r }
            if b == i { break }
            a.swapAt(i, b)
            i = b
        }
    }
    return top
}

var heap: [Int] = []
for v in [5, 3, 8, 1, 4] { pushHeap(&heap, v) }
var previous = Int.max
while !heap.isEmpty {
    let x = popHeap(&heap)
    if x > previous { fatalError("not a max-heap order") }
    previous = x
}
print("max heap ok")
